dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/net/DataOkHttpUploaderV2.kt

```diff
@@ -0,0 +1,162 @@
+
+    private val userAgent by lazy {
+        System.getProperty(DataOkHttpUploader.SYSTEM_UA).let {
+            if (it.isNullOrBlank()) {
+                "Datadog/${BuildConfig.SDK_VERSION_NAME} " +
+                    "(Linux; U; Android ${Build.VERSION.RELEASE}; " +
+                    "${Build.MODEL} Build/${Build.ID})"
+            } else {
+                it
+            }
+        }
+    }
+
+    // region DataUploader
+
+    @Suppress("TooGenericExceptionCaught")
+    override fun upload(data: ByteArray): UploadStatus {
+        val requestId = UUID.randomUUID().toString()
+        val uploadStatus = try {
+            executeUploadRequest(data, requestId)
+        } catch (e: Throwable) {
+            internalLogger.e("Unable to upload batch data.", e)
+            UploadStatus.NETWORK_ERROR
+        }
+
+        uploadStatus.logStatus(uploaderName, data.size, devLogger, false, requestId)
+        uploadStatus.logStatus(uploaderName, data.size, internalLogger, true, requestId)
+
+        return uploadStatus
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun executeUploadRequest(
+        data: ByteArray,
+        requestId: String
+    ): UploadStatus {
+        val request = buildRequest(data, requestId)
+        val call = callFactory.newCall(request)
+        val response = call.execute()
+        return responseCodeToUploadStatus(response.code())
+    }
+
+    private fun buildRequest(data: ByteArray, requestId: String?): Request {
+        val builder = Request.Builder()
+            .url(buildUrl())
+            .post(RequestBody.create(null, data))
+
+        buildHeaders(builder, requestId)
+
+        return builder.build()
+    }
+
+    private fun buildUrl(): String {
+        val queryParams = buildQueryParameters()
+        return if (queryParams.isEmpty()) {
+            intakeUrl
+        } else {
+            intakeUrl + queryParams.map { "${it.key}=${it.value}" }.joinToString("&", prefix = "?")
+        }
+    }
+
+    private fun buildHeaders(builder: Request.Builder, requestId: String?) {
+        builder.addHeader(HEADER_API_KEY, clientToken)
+        builder.addHeader(HEADER_EVP_ORIGIN, source)
+        builder.addHeader(HEADER_EVP_ORIGIN_VERSION, BuildConfig.SDK_VERSION_NAME)
+        builder.addHeader(HEADER_USER_AGENT, userAgent)
+        builder.addHeader(HEADER_CONTENT_TYPE, contentType)
+
+        if (!requestId.isNullOrBlank()) {
+            builder.addHeader(HEADER_REQUEST_ID, requestId)
+        }
+    }
+
+    protected open fun buildQueryParameters(): Map<String, Any> {
+        return emptyMap()
+    }
+
+    private fun responseCodeToUploadStatus(code: Int): UploadStatus {
+        return when (code) {
+            HTTP_ACCEPTED -> UploadStatus.SUCCESS
+            HTTP_BAD_REQUEST -> UploadStatus.HTTP_CLIENT_ERROR
+            HTTP_UNAUTHORIZED -> UploadStatus.INVALID_TOKEN_ERROR
+            HTTP_FORBIDDEN -> UploadStatus.HTTP_CLIENT_ERROR
```

Comment:
From my tests, when the client token is invalid (e.g. misspelled), we're receiving `403` (forbidden). Why do we map `401` (unauthorized) to `INVALID_TOKEN_ERROR` here? Do we mean the case where token was valid but was revoked?

Reply:
I see 👌, it seems fine then. My only concern was the log message here:
```kotlin
INVALID_TOKEN_ERROR -> if (!ignoreInfo) {
   logger.e(
      "$batchInfo failed because your token is invalid. " +
      "Make sure that the provided token still exists."
   )
}
```
but from the spec you quoted it totally makes sense to only print it on `401` and consider `403` as client error 👍. I will do this update in my PR, as I did the opposite...